File: albedo-math/src/main/java/jp/albedo/jpl/files/util/LittleEndianByteBufferReader.java
package jp.albedo.jpl.files.util;

import java.nio.ByteBuffer;

import static java.nio.ByteOrder.LITTLE_ENDIAN;

public class LittleEndianByteBufferReader {

    private final ByteBuffer byteBuffer;

    private final byte[] intBuffer = new byte[4];

    private final byte[] doubleBuffer = new byte[8];

    private final byte[] stringBugger = new byte[100];

    public LittleEndianByteBufferReader(ByteBuffer byteBuffer) {
        this.byteBuffer = byteBuffer;
    }

    public int getInt() {
        byteBuffer.get(intBuffer);
        return ByteBuffer.wrap(intBuffer).order(LITTLE_ENDIAN).getInt();
    }

    public double getDouble() {
        byteBuffer.get(doubleBuffer);
        return ByteBuffer.wrap(doubleBuffer).order(LITTLE_ENDIAN).getDouble();
    }

    public double[] readDoubles(int n) {
        double[] doubles = new double[n];
        for (int i = 0; i < n; i++) {
            byteBuffer.get(doubleBuffer);
            doubles[i] = ByteBuffer.wrap(doubleBuffer).order(LITTLE_ENDIAN).getDouble();
        }
        return doubles;
    }

    public String getString(int n) {
        if (n > 100) {
            throw new UnsupportedOperationException("Buffer too small to read " + n + " chars!");
        }

        byteBuffer.get(stringBugger, 0, n);
        return new String(stringBugger);
    }

}

File: api/jpl/files/LittleEndianByteBufferReader.ts
export class LittleEndianByteBufferReader {

    private dataView: DataView;
    private index: number = 0;

    constructor(byteBuffer: Uint8Array) {
        this.dataView = new DataView(byteBuffer.buffer);
    }

    getInt(): number {
        if (this.index + 4 > this.dataView.byteLength) {
            throw Error('Out of index error');
        }
        const value = this.dataView.getUint32(this.index, true);
        this.index += 4;
        return value;
    }

    getDouble(): number {
        if (this.index + 8 > this.dataView.byteLength) {
            throw Error('Out of index error');
        }
        const value = this.dataView.getFloat64(this.index, true);
        this.index += 8;
        return value;
    }

    readDoubles(n: number): number[] {
        if (this.index + n * 8 > this.dataView.byteLength) {
            throw Error('Out of index error');
        }
        const doubles = new Array<number>(n);
        for (let i = 0; i < n; i++) {
            doubles[i] = this.getDouble();
        }
        return doubles;
    }

    getString(n: number): string {
        if (this.index + n > this.dataView.byteLength) {
            throw Error('Out of index error');
        }
        var result = '';
        for (let i = 0; i < n; i++) {
            result += String.fromCharCode(this.dataView.getUint8(this.index + i));
        }
        this.index += n;
        return result;
    }

}
